File: ELA.CR.NewProposal/core/transaction/payload/Record.go
package payload

import (
	"errors"
	"io"

	"Elastos.ELA/common/serialize"
)

const RecordPayloadVersion byte = 0x00

type Record struct {
	RecordType string
	RecordData []byte
}

func (a *Record) Data(version byte) []byte {
	//TODO: implement RegisterRecord.Data()
	return []byte{0}
}

// Serialize is the implement of SignableData interface.
func (a *Record) Serialize(w io.Writer, version byte) error {
	err := serialize.WriteVarString(w, a.RecordType)
	if err != nil {
		return errors.New("[RecordDetail], RecordType serialize failed.")
	}
	err = serialize.WriteVarBytes(w, a.RecordData)
	if err != nil {
		return errors.New("[RecordDetail], RecordData serialize failed.")
	}
	return nil
}

// Deserialize is the implement of SignableData interface.
func (a *Record) Deserialize(r io.Reader, version byte) error {
	var err error
	a.RecordType, err = serialize.ReadVarString(r)
	if err != nil {
		return errors.New("[RecordDetail], RecordType deserialize failed.")
	}
	a.RecordData, err = serialize.ReadVarBytes(r)
	if err != nil {
		return errors.New("[RecordDetail], RecordData deserialize failed.")
	}
	return nil
}

File: SPV.CPP/SDK/Plugin/Transaction/Payload/Record.cpp
// Copyright (c) 2012-2018 The Elastos Open Source Project
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

#include "Record.h"

#include <Common/Log.h>
#include <Common/Utils.h>

namespace Elastos {
	namespace ElaWallet {

		Record::Record() :
				_recordType("") {

		}

		Record::Record(const std::string &recordType, const bytes_t &recordData) {
			_recordType = recordType;
			_recordData = recordData;
		}

		Record::Record(const Record &payload) {
			operator=(payload);
		}

		Record::~Record() {
		}

		void Record::SetRecordType(const std::string &recordType) {
			_recordType = recordType;
		}

		void Record::SetRecordData(const bytes_t &recordData) {
			_recordData = recordData;
		}

		std::string Record::GetRecordType() const {
			return _recordType;
		}

		bytes_t Record::GetRecordData() const {
			return _recordData;
		}

		size_t Record::EstimateSize(uint8_t version) const {
			size_t size = 0;
			ByteStream stream;

			size += stream.WriteVarUint(_recordType.size());
			size += _recordType.size();
			size += stream.WriteVarUint(_recordData.size());
			size += _recordData.size();

			return size;
		}

		void Record::Serialize(ByteStream &ostream, uint8_t version) const {
			ostream.WriteVarString(_recordType);
			ostream.WriteVarBytes(_recordData);
		}

		bool Record::Deserialize(const ByteStream &istream, uint8_t version) {
			if (!istream.ReadVarString(_recordType)) {
				Log::error("Payload record deserialize type fail");
				return false;
			}

			if (!istream.ReadVarBytes(_recordData)) {
				Log::error("Payload record deserialize data fail");
				return false;
			}

			return true;
		}

		nlohmann::json Record::ToJson(uint8_t version) const {
			nlohmann::json j;

			j["RecordType"] = _recordType;
			j["RecordData"] = _recordData.getHex();

			return j;
		}

		void Record::FromJson(const nlohmann::json &j, uint8_t version) {
			_recordType = j["RecordType"].get<std::string>();
			_recordData.setHex(j["RecordData"].get<std::string>());
		}

		IPayload &Record::operator=(const IPayload &payload) {
			try {
				const Record &payloadRecord = dynamic_cast<const Record &>(payload);
				operator=(payloadRecord);
			} catch (const std::bad_cast &e) {
				Log::error("payload is not instance of Record");
			}

			return *this;
		}

		Record &Record::operator=(const Record &payload) {
			_recordData = payload._recordData;
			_recordType = payload._recordType;

			return *this;
		}

	}
}